### DISCUSSION TITLE
How to place a .txt file and a .tiff file onto a PDF and get a new PDF

### DISCUSSION THREAD

**Discussion by 74Bravo-x** [ImageMagick/ImageMagick#4046] [Help]
Hello all.  I am fairly new to Ubuntu (using version 18.04) and very new to ImageMagick (version 6.9.7-4 Q16 x86_64).

I have a .txt and a .tiff that I would like to place on a PDF.  The PDF has boxes and lines and such on it (think of a service or parts invoice), so I need to be able to specify the position of the .txt and .tiff on the PDF.  The end result would be a new PDF with the info from .txt and .tiff placed on it.

I am thinking that convert could work for this.  But I am not sure how to handle the positioning of the .txt and .tiff. 

Oh yeah, this all needs to be done by the command line!

Is convert a good way to do this, or are there better ways?

Thanks!

**Comment by fmw42** [comment]
Convert can do the job. For the image, you could use

convert image.pdf image.tif[0] -geometry +X+Y -compose over -composite result.pdf

But note that ImageMagick rasterizes you input PDF and the output PDF will then be a raster image embedded in a vector shell. So if your input PDF is vector, you will lose the vectors and have raster pixels.

For the text file, you can create a new image using a number of different methods, but caption: is likely going to be what you want. Then do the above with the text image that you just created.

See

https://legacy.imagemagick.org/Usage/compose/
https://legacy.imagemagick.org/Usage/layers/#convert
https://legacy.imagemagick.org/Usage/text/#caption

**Reply by Jmuccigr** [reply]
As @fmw42 notes, this is going to mess with your original PDF and possibly blow up the size of it. If you want to keep it, I'd suggest using IM to combine the text and tiff into one file, either a pdf directly or another image format that you can then use img2pdf to turn that into a pdf of the appropriate size, finally qpdf or pdftk to overlay that new pdf onto the original.

**Comment by 74Bravo-x** [comment]
Thanks for the replies.  I got put on another project all of the sudden but now I am back to Imagemagick.  Here goes!

**Comment by 74Bravo-x** [comment]
So for now I am going to ignore the .tif.  I am just trying to place the contents of the .txt file onto the .pdf.

I tried

$ convert new.txt j1137C.pdf new2.pdf

and got a few errors and no output file-

convert-im6.q16: improper image header `new.txt' @ error/txt.c/ReadTXTImage/443.
convert-im6.q16: not authorized `j1137C.pdf' @ error/constitute.c/ReadImage/412.
convert-im6.q16: no images defined `new2.pdf' @ error/convert.c/ConvertImageCommand/3258

In theory, should this work?

Thanks

**Reply by fmw42** [reply]
You need to convert your text into an image using label:@filename.txt or caption:@filename.txt  Then composite or append that into your PDF

See https://legacy.imagemagick.org/Usage/text/

**Comment by snibgo** [comment]
> im6.q16: not authorized j1137C.pdf' @ error/constitute.c/ReadImage/412.

This is also a problem. I suspect your policy.xml prohibits you from reading PDF files.

**Reply by 74Bravo-x** [reply]
> > im6.q16: not authorized j1137C.pdf' @ error/constitute.c/ReadImage/412.
> 
> This is also a problem. I suspect your policy.xml prohibits you from reading PDF files.

Nice catch.  Did an edit on policy.xml and the error is gone and I can read PDFs now.  Thank you!

**Comment by 74Bravo-x** [comment]
So I read up on label and caption.  The .txt file I am trying to place on the pdf has 93 words in it, which looks to be too much.  Is there another method I could use?

**Comment by fmw42** [comment]
Caption will work. It it can reduce the font size automatically if you specify the width and height that you want.

**Comment by 74Bravo-x** [comment]
convert label:@actual.txt labelfile.gif
 
gave me this message-

convert-im6.q16: not authorized `@actual.txt' @ error/property.c/InterpretImageProperties/3516.

Like what happened above with getting the pdf not authorized message, I went to policy.xml but didn't see anything preventing me from opening .txt files?

**Reply by fmw42** [reply]
Look for the policy about @ and relax it.

**Reply by 74Bravo-x** [reply]
Thanks, can't believe I missed the @ part.  "Relax it"  well said, will do.  Thanks

**Comment by 74Bravo-x** [comment]
It worked, ran this command with no errors or issues-

convert label:@actual.txt labelfile.gif

So now the .txt file has been converted to a .gif.  Now it should just be a matter of placing the .gif onto the .pdf?

**Comment by 74Bravo-x** [comment]
convert label:@actual.txt labelfile.gif

It worked!  Though going from the .txt to .gif resulted in the text spacing being off, but I can look into that later.

So I ran 
convert labelfile.gif j1137C.pdf newpdf.pdf

And instead of placing the text on top of the pdf, it took my one page pdf and just added the text to a second page of the pdf.

Am I barking up the wrong tree here?  Never thought it would be such a journey to place a text file on top of a pdf!

**Comment by snibgo** [comment]
I suppose your input PDF has a number of pages. When you read that file, IM creates a list of images. If you then read your GIF file, this adds another image to the list. (Or, as you have done, in the opposite order.) So you have a list of images. When you write those images to a PDF, that creates a PDF with one page per image.

Your OP is vague about what result you want. Perhaps you want to composite your text over every PDF page. For that, your text image needs to have a transparent background ("None"), and "-layers composite" can combine them. See https://imagemagick.org/script/command-line-options.php#layers . Something like this:

```
magick in.pdf NULL: -background None label:@actual.txt -layers composite out.pdf
```

**Comment by 74Bravo-x** [comment]
It's a one page .txt being placed on top of a one page pdf.  Thanks for the advice on the transparent background!  Will give that a try.

**Comment by fmw42** [comment]
If you have a multi-page PDF, then which page do you want the text placed and where on the page?

**Comment by 74Bravo-x** [comment]
YES, made some progress!  I ran this command-
convert j1137C.pdf NULL: -background None label:@actual.txt -layers composite out.pdf

And it placed the text on the pdf!  

![Captureupload](https://user-images.githubusercontent.com/88505561/135136363-75e00d82-431b-409a-a6cf-361c4aba021e.PNG)

The lines, boxes, and fine print at the bottom are all part of the pdf.  The .txt part is make, model, sign up for special discounts, taxable, parts non taxable and so on.  Of course I need to work on positioning the txt properly on the pdf so everything lines up.  Looks like the gravity command might work for that?

**Comment by 74Bravo-x** [comment]
After reading up on it, looks like gravity is going to be a fun one.  It works with label, so no problems there.  

But I am having a hard time figuring out where to insert "-gravity Center" within the command?

convert j1137C.pdf NULL: -background None label:@actual.txt -layers composite out.pdf

**Reply by fmw42** [reply]
You can place it right before label:

**Comment by snibgo** [comment]
"-gravity" affects "-layers composite", so it can be anywhere before "-layers composite". If you put it after "-layers composite", it will have no effect.

**Comment by 74Bravo-x** [comment]
Now that is a beautiful thing!  Thanks to both of you.

**Comment by 74Bravo-x** [comment]
convert j1137C.pdf NULL: -background None -geometry -45-8 -gravity Center label:@EXAM2.TXT -layers composite outp13.pdf

So this command works well, and this part of the command "-geometry -45-8 -gravity Center" places the text on top of the pdf _almost_ where I want it.  More tinkering to do there, but I get how geometry works.

But I noticed something odd about the text in the output pdf.

Here is a shot of the original .txt file.  See how the numbers all line up?
![snip01](https://user-images.githubusercontent.com/88505561/135675497-e9d91c0e-ef95-4a5e-b033-8d935b7973f4.PNG)

But after I run the convert command, the numbers don't line up-
![snip02](https://user-images.githubusercontent.com/88505561/135675959-2da8b203-bd9f-46c0-8a69-97739357c587.PNG)

Is there a way I can make sure that convert leaves the original text alone?  Like no changes to spacing and such?

**Comment by snibgo** [comment]
Your first example uses a monospaced font. This means all glyphs (including the space character) take the same width. The font in your second example is not monospaced.

So the cure is to use a monospaced font. One example is Courier.

**Comment by fmw42** [comment]
Create the text with gravity west. But composite it with gravity center. Does this help?

`convert j1137C.pdf NULL: -background None label:@EXAM2.TXT -geometry -45-8 -gravity Center -layers composite outp13.pdf`

**Comment by 74Bravo-x** [comment]
Monospaced font, that sounds good.  I can use Courier New, which looks to be a monospaced font.

So I went back to the old HP-UX 11.11 server (where the text is generated by a program), grabbed a fresh copy of the text, then made sure the resulting .txt file was Courier New.  Placed the .txt file on the Ubuntu server and got the same thing.

The text file

![text01](https://user-images.githubusercontent.com/88505561/136260905-0c252f24-b65d-4e0a-b149-8bd5d3b1dd10.PNG)

And the end result-

![pdf01](https://user-images.githubusercontent.com/88505561/136261088-02daf6a5-293d-4d5a-9dde-08e29cab8c62.PNG)

**Comment by snibgo** [comment]
> ... then made sure the resulting .txt file was Courier New.

Text files contain just text. They do not contain metadata about fonts.

In your "end result", the letter I is clearly not as wide as the other letters. You are not using a monospaced font.

What is your IM command? In particular, what is your "-font" option?

**Comment by 74Bravo-x** [comment]
This is the latest version-

convert j1137C.pdf NULL: -background None -geometry -45-8 -gravity Center label:@text2.txt -layers composite outp17.pdf

Will see what font options are available in IM.

**Comment by snibgo** [comment]
You haven't used a "-font" option, so IM used a default font that isn't monospaced.

"convert -list font" will list available fonts.

**Comment by 74Bravo-x** [comment]
Ok, thanks to your suggestions I added font info to the command-

convert j1137C.pdf NULL: -background None -font Courier -pointsize 11 -geometry -45-42 -gravity Center label:@text2.txt -layers composite outp30.pdf

And that helped a great deal.  But I am still seeing some strange stuff.  Like here, everything is aligned on the decimal-
![Capturefont](https://user-images.githubusercontent.com/88505561/136819034-f3105c1f-a9e5-4423-835c-b0a2b8e07b66.PNG)
But the words still got shifted around somehow?

**Comment by snibgo** [comment]
Please paste a link to text2.txt, or a zip of that file. I need to see the actual file, not a copy-and-paste version of it. Perhaps it contains tab characters or something.

**Reply by 74Bravo-x** [reply]
Here is text2.txt-

https://www.mediafire.com/file/3udxu6o3v2dnp1c/text2.txt/file

**Comment by snibgo** [comment]
That website requires me to agree that they can steal my data before I can download stuff. I don't.

**Comment by 74Bravo-x** [comment]
Sorry about that.  I haven't done any uploading since Photobucket went from free to $400 a few years back!

Let's see if this works

[text2.txt](https://github.com/ImageMagick/ImageMagick/files/7324742/text2.txt)

**Comment by snibgo** [comment]
You have "-geometry -45-42 -gravity Center" before reading and rasterizing the text file. "-gravity Center" affects this rasterizing, and you don't want that. So move "-geometry -45-42 -gravity Center" to after reading the text file.

**Comment by 74Bravo-x** [comment]
Thank you, this worked!!

$convert j1137C.pdf NULL: -background None -font Courier -pointsize 11 label:@text2.txt -geometry -45-42 -gravity Center -layers composite outp32.pdf

But this is an odd one.  When I align this north / south, the DUPLICATE DUPLICATE text should sit right on top of that line on the underlying pdf.  And it does!  But the MAKE, MODEL, SERNO section needs to come down a little.  The PARTS NONTAXABLE needs to come down alot.  The 68.30 needs to be down to PLEASE PAY THIS TOTAL.

Is it even possible to further manipulate the contents of the text to that degree?
![Capture1012](https://user-images.githubusercontent.com/88505561/137166525-d9cbb102-57d9-45f8-9151-3568d52e97b1.PNG)

**Comment by snibgo** [comment]
In your text file, inset two blank lines before "MAKE:", and whatever number of blank lines you need before PARTS NONTAXABLE etc.

**Comment by 74Bravo-x** [comment]
I see what you are saying.  Open the text file in an editor, set the spacing properly by hand, then place it on the pdf.  Which would work, but I need to be able to do this all from the command line.  Maybe I've finally hit a wall here.  So close though.

**Comment by snibgo** [comment]
Unix has many utilities for manipulating text files. Doubtless you can insert blank lines from the command line.

But where does that text file come from? I suppose it is automatically created, perhaps using some templates. The templates need more blank lines.
